Create work packages on IFC model page

ulferts opened the pull request:
Adds the capability to create work_packages from the ifc_models page. The L&F is similar to how work packages are created on the work package list page.

## TODOs

* [x] Check permissions to disable the button if the `add_work_packages` permission is lacking.
* [x] ~~Add the ability to add BCFs to the `ifc_models/:id` page as well.~~ No longer necessary since route refactoring
* [x] Add the split create on the `list-only` mode as well
* [x] Generalize the split screen create for work package and bim using the route template
* [x] Allow creating subsequent BCFs. Currently, the button is disabled.
* [x] Styling so that e.g. the lower right save bar does not have a bottom margin. 




## Out of scope

* Add a viewpoint to the created work package
  * Error handling (transaction) when one of the requests fail. We need to remove the work package again in such cases but might not be allowed to.
* Decide on additional places where BCFs are to be created. 
*  Different success messages upon creation. Currently, a "The work package has been successfully created" message is displayed.

https://community.openproject.com/projects/openproject/work_packages/30519

ulferts committed:
round trip for creating a work package on ifc page

openprojectci commented:
<!--
  0 Errors
  1 Warning: <a href="https://github.com/op...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/150941977b158758cd4d8027ba77f28223d2fe25/frontend/src/app/components/modals/editor/macro-wp-button-modal/wp-button-macro.modal.ts#L37">frontend/src/app/components/modals/editor/macro-wp-button-modal/wp-button-macro.modal.ts#L37</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ulferts committed:
fix aborting bcf creation

ulferts committed:
check create permission

openprojectci commented on frontend/src/app/components/wp-new/wp-new-full-view.component.ts:
```diff
@@ -26,14 +26,14 @@
-import {WorkPackageCreateController} from 'core-components/wp-new/wp-create.controller';
+import {WorkPackageCreateComponent} from 'core-components/wp-new/wp-create.component';
 import {Component} from '@angular/core';
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/8bc9e1f18f1494c2afdb99f865cf858298ae755e/frontend/src/app/components/wp-new/wp-new-full-view.component.ts#L31"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-new/wp-new-split-view.component.ts:
```diff
@@ -26,13 +26,13 @@
-import {WorkPackageCreateController} from 'core-components/wp-new/wp-create.controller';
+import {WorkPackageCreateComponent} from 'core-components/wp-new/wp-create.component';
 import {Component} from '@angular/core';
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/8bc9e1f18f1494c2afdb99f865cf858298ae755e/frontend/src/app/components/wp-new/wp-new-split-view.component.ts#L31"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into bim/feature/create_bcf

oliverguenther committed:
Avoid calling cdRef when destroyed

oliverguenther committed:
Add create button to partitioned toolbar

oliverguenther committed:
Fix routes for split create

oliverguenther committed:
Update AuthorisationService when creating new work package

Otherwise, the information about the new work package's permissions are missing

oliverguenther committed:
Rename route in split screen

oliverguenther committed:
Add ChangeDetection.onPush where requested

oliverguenther committed:
Make new split view use generic route template

oliverguenther committed:
Specs

openprojectci commented:
<!--
  0 Errors
  1 Warning: <a href="https://github.com/op...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/opf/openproject/blob/081ba59455a4ea64fa1f7c7f1e0a8120f02c3ee4/frontend/src/app/components/modals/editor/macro-wp-button-modal/wp-button-macro.modal.ts#L45">frontend/src/app/components/modals/editor/macro-wp-button-modal/wp-button-macro.modal.ts#L45</a> - Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

HDinger reviewed (changes requested):
All in one it works. But I found some issues:

* When the details split screen is open, a click on the create button does nothing (on the ifc page).
* When starting on the ifc defaults page, there is a border shown around the card view and the spacing to the right is missing. Both will vanish, when going to the details view of a WP and back again to the card view. I assume that a class is not correctly set.
<img width="809" alt="Bildschirmfoto 2020-03-04 um 10 16 05" src="https://user-images.githubusercontent.com/7457313/75864021-af76d400-5e01-11ea-94a1-4accfad32792.png">

* A similar behaviour happens on the WP page: After opening the split screen and closing it again, there is too much spacing to the right:
<img width="720" alt="Bildschirmfoto 2020-03-04 um 10 14 48" src="https://user-images.githubusercontent.com/7457313/75864205-f2d14280-5e01-11ea-86dd-ff06f0963430.png">

* When starting on `/work_packages` and opening the create form, the button is not disabled:
<img width="721" alt="Bildschirmfoto 2020-03-04 um 10 15 08" src="https://user-images.githubusercontent.com/7457313/75864297-14cac500-5e02-11ea-9508-9021d5e7dcaa.png">
If you then reload the page and close the create form, the button stays disabled:
<img width="732" alt="Bildschirmfoto 2020-03-04 um 10 15 23" src="https://user-images.githubusercontent.com/7457313/75864364-2a3fef00-5e02-11ea-9640-cbc325973d76.png">
The same happens on the ifc page.

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into bim/feature/create_bcf

oliverguenther committed:
Use the correct new route given the current route

Depending on what route we're in, we need to open a different route
for the new state. In the details view, we cannot use `.new` but would
need to use `^.new`. To avoid confusion, we define the route for new
split view in the routes itself.

oliverguenther committed:
Use separate class for the details view

the split class and details class were using the same class name

oliverguenther committed:
Correctly update disabled state in wp-create-button

oliverguenther commented:
@HDinger I think I caught all of them

HDinger reviewed (commented):
@oliverguenther One last thing: When I am on `projects/foo/bcf/list?models...` and click on create, the state changes to the viewer and list. I guess it should stay in the list only with an opened create form.

oliverguenther committed:
Update new state name when route changes

oliverguenther committed:
Extend create spec for all cases

HDinger reviewed (approved):
👍 Good to merge once travis is ready

oliverguenther committed:
Fix newRoute on full view

openprojectci commented on frontend/src/app/modules/work_packages/routing/wp-full-view/wp-full-view.component.ts:
```diff
@@ -34,6 +34,7 @@ import {Component, Injector, OnInit} from '@angular/core';
 import {WorkPackageSingleViewBase} from "core-app/modules/work_packages/routing/wp-view-base/work-package-single-view.base";
 import {BackRoutingService} from "core-app/modules/common/back-routing/back-routing.service";
+import {of} from "rxjs";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/opf/openproject/blob/081ba59455a4ea64fa1f7c7f1e0a8120f02c3ee4/frontend/src/app/modules/work_packages/routing/wp-full-view/wp-full-view.component.ts#L38"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</td>
    </tr>
  </tbody>
</table>
